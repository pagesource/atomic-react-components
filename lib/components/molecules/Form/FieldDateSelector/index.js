// @flow
export { default } from './FieldDateSelector';
export { FieldDateSelectorVanilla } from './FieldDateSelector';
