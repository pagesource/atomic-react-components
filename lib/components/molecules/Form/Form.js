// @flow
import React from 'react';
import { useForm } from 'react-hook-form';
import type { Node } from 'react';
import { buildSchema } from '../../../core/utils/buildValidationSchema';
import Debug from './Debug';
import FieldInput from './FieldInput';
import FieldSelect from './FieldSelect';
import FieldTextarea from './FieldTextarea';
import FieldRadioGroup from './FieldRadioGroup';
import FieldCheckbox from './FieldCheckbox';
import FieldToggle from './FieldToggle';
import FieldFileInput from './FieldFileInput';
import DDFileInput from './DragDropInput';

type Props = {
  initialValues?: any,
  children: Node,
  onSubmit: Function,
  debug?: boolean,
  validationSchema?: Object | null,
  validate?: String,
  yupValidationSchema?: Function | null,
};

const Form = ({
  onSubmit,
  initialValues,
  children,
  debug,
  validate,
  validationSchema,
  yupValidationSchema,
}: Props) => {
  const getSchema = () => {
    if (yupValidationSchema) return yupValidationSchema;
    return validationSchema ? buildSchema(validationSchema) : null;
  };
  const { handleSubmit, register, errors, setValue } = useForm({
    mode: validate,
    validationSchema: getSchema(),
    defaultValues: initialValues,
  });

  // $flow-disable-line  eslint-disable-line
  const childrenWithProps = children[0].props.children.map((child) => {
    return React.cloneElement(child, { register, errors, setValue });
  });

  // $flow-disable-line  eslint-disable-line
  childrenWithProps.push(children[1]);
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {childrenWithProps}
      {debug && <Debug />}
    </form>
  );
};

Form.Checkbox = FieldCheckbox;

Form.Input = FieldInput;

Form.RadioGroup = FieldRadioGroup;

Form.Select = FieldSelect;

Form.Textarea = FieldTextarea;

Form.Toggle = FieldToggle;

Form.FieldFile = FieldFileInput;

Form.DDFile = DDFileInput;

Form.defaultProps = {
  debug: false,
  initialValues: null,
  validationSchema: null,
  yupValidationSchema: null,
  validate: 'onSubmit',
};

export default Form;
