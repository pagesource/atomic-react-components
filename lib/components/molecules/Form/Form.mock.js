import { action } from '@storybook/addon-actions';

const userFormProps = {
  initialValues: {
    email: '',
    username: '',
    typeOfUser: 'Admin',
    comment: '',
    ageGroup: '',
    termsAndConditions: false,
  },
  debug: false,
  action: '/',
  validationSchema: {
    username: {
      max: [15, 'Max limit reached'],
      required: 'User name Required',
    },
    email: {
      max: [30, 'Max reached'],
      email: 'Invalid email format',
      required: 'Required',
    },
    comment: {
      max: [100, 'Comment should not be more than 100 characters'],
      required: 'Please enter a comment',
    },
    ageGroup: {
      required: 'Please select an Age group',
    },
    termsAndConditions: {
      type: 'bool',
      test: [
        'termsAndConditions',
        'please accept terms and conditions!!!',
        (value) => value === true,
      ],
    },
  },
  onSubmit: action('On Submit Fired'),
};

const userFormSelectProps = {
  options: ['Admin', 'Read Only'],
  name: 'typeOfUser',
  label: 'Type of user',
};

const userFormNameInputProps = {
  name: 'username',
  type: 'text',
  label: 'User Name',
  validationInstructions: 'Enter valid username',
};

const fileMultipleInputProps = {
  name: 'fileuploadMulti',
  localeText: {
    btnTextPrimary: 'Choose Files',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop Files Here...',
  },
  multiple: true,
};
const fileSingleInputProps = {
  name: 'fileupload',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};
const fileDDInputProps = {
  name: 'fileuploadDD',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};

const userFormEmailProps = {
  type: 'email',
  name: 'email',
  label: 'Email',
  validationInstructions: 'Enter valid email',
};

const userFormCommentProps = {
  name: 'comment',
  label: 'Comment',
  placeholder: 'Enter your comment here...',
};

const userFormAgeGroupProps = {
  name: 'ageGroup',
  radioGroupArray: [
    {
      value: '30-',
      label: '< 30',
    },
    {
      value: '30-60',
      label: '30-60',
    },
    {
      value: '60+',
      label: '> 60',
    },
  ],
};

const userFormTnCProps = {
  name: 'termsAndConditions',
  label: 'I accept terms & conditions',
};

const userFormToggleProps = {
  name: 'test_1',
  id: 'checkbox_test_1',
  value: 'off',
  label: 'Toggle',
  indicatorOn: 'On',
  indicatorOff: 'Off',
  disabled: false,
  intialToggleState: false,
  displayIndicators: true,
};

export {
  userFormProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormSelectProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormTnCProps,
  userFormToggleProps,
  fileDDInputProps,
};
