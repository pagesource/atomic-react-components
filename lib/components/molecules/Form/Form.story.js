// @flow
import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  userFormProps,
  userFormSelectProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormDateSelectorProps,
  userFormTnCProps,
  userFormToggleProps,
} from './Form.mock';

// Import Vanilla Component to display markup and props
import Form from './Form';
import Button from '../../atoms/Button';
// Import Styled Component to showcase variations

storiesOf('Molecules/Form', module)
  .addParameters({ jest: ['Form', 'Button'] })
  .add('Default', () => (
    <Form {...userFormProps}>
      <div>
        <Form.Select {...userFormSelectProps} />
        <Form.Input {...userFormNameInputProps} />
        <Form.Input {...userFormEmailProps} />
        <Form.FieldFile {...fileMultipleInputProps} />
        <Form.FieldFile {...fileSingleInputProps} />
        <Form.DDFile {...fileSingleInputProps} />
        <Form.RadioGroup {...userFormAgeGroupProps} />
        <Form.DateSelector {...userFormDateSelectorProps} />
        <Form.Checkbox {...userFormTnCProps} />
        <Form.Textarea {...userFormCommentProps} />
        <Form.Toggle {...userFormToggleProps} />
      </div>
      <Button primary type="submit">
        Submit
      </Button>
    </Form>
  ));
