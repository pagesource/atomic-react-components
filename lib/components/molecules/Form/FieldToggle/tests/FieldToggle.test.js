import React from 'react';
import { FieldToggleVanilla } from '../index';
import { userFormToggleProps } from '../../Form.mock';
import { shallowWithTheme, mountWithTheme } from '../../../../../../__mock__/themeMock';

describe('<FieldToggle />', () => {
  let FieldToggleComponent = '';

  test('should render correctly', () => {
    FieldToggleComponent = shallowWithTheme(<FieldToggleVanilla {...userFormToggleProps} />);
    expect(FieldToggleComponent).toMatchSnapshot();
  });

  test('should render label correctly', () => {
    FieldToggleComponent = mountWithTheme(<FieldToggleVanilla {...userFormToggleProps} />);
    expect(FieldToggleComponent.find('label').length).toEqual(2);
  });

  test('should render label with Value correctly', () => {
    FieldToggleComponent = mountWithTheme(<FieldToggleVanilla {...userFormToggleProps} />);
    expect(
      FieldToggleComponent.find('.label-name')
        .at(0)
        .text()
    ).toEqual(userFormToggleProps.label);
  });

  test('should render input correctly', () => {
    FieldToggleComponent = mountWithTheme(<FieldToggleVanilla {...userFormToggleProps} />);
    expect(FieldToggleComponent.find('input').length).toEqual(1);
  });

  test('should render disabled input', () => {
    FieldToggleComponent = mountWithTheme(<FieldToggleVanilla {...userFormToggleProps} disabled />);
    expect(FieldToggleComponent.find('input').prop('disabled')).toEqual(true);
  });

  test('should tigger click event with toggle Value', () => {
    const handleToggle = jest.fn();
    jest.spyOn(React, 'useState').mockImplementation(toggle => [toggle, handleToggle]);
    FieldToggleComponent = mountWithTheme(<FieldToggleVanilla {...userFormToggleProps} />);
    FieldToggleComponent.find('input[type="checkbox"]').simulate('click', {
      target: { clicked: true },
    });
    expect(handleToggle).toHaveBeenCalledWith(true);
  });

  test('should render Indicator On', () => {
    FieldToggleComponent = mountWithTheme(
      <FieldToggleVanilla {...userFormToggleProps} intialToggleState />
    );
    FieldToggleComponent.find('input[type="checkbox"]').simulate('click', { target: {} });
    expect(FieldToggleComponent.find('.toggle-label').text()).toEqual(
      userFormToggleProps.indicatorOn
    );
  });

  test('should tigger a handleToggleCallback function', () => {
    const handleToggleCallback = jest.fn();
    FieldToggleComponent = mountWithTheme(
      <FieldToggleVanilla {...userFormToggleProps} handleToggleCallback={handleToggleCallback} />
    );
    FieldToggleComponent.find('input[type="checkbox"]').simulate('click', {
      target: { clicked: true },
    });
    expect(handleToggleCallback).toHaveBeenCalled();
  });
});
