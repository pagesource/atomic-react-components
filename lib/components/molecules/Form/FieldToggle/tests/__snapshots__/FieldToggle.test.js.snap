// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldToggle /> should render correctly 1`] = `
<div
  className="toggle-container"
>
  <Label
    className="label-name"
    elementLocator=""
    htmlFor="checkbox_test_1"
  >
    Toggle
  </Label>
  <div>
    <Label
      className=""
      elementLocator=""
      htmlFor="checkbox_test_1"
    >
      <Input__StyledInput
        className="toggle-input"
        disabled={false}
        id="checkbox_test_1"
        name="test_1"
        onClick={[Function]}
        type="checkbox"
        value="off"
      />
      <span
        className="slider"
      />
    </Label>
    <span
      className="toggle-label"
    >
      Off
    </span>
  </div>
</div>
`;
